api_spec.yaml

```diff
@@ -1377,7 +1378,140 @@ paths:
         
         '500':
           $ref: '#/components/responses/500'
-  
+
+  /recipe/userMetadata:
+    get:
+      tags:
+        - User Metadata Recipe
+      operationId: userMetadataRead
+      description: |
+        Gets the stored metadata object of the user
+      parameters:
+        - $ref: '#/components/parameters/userMetadataRid'
+        - $ref: '#/components/parameters/api-key'
+        - $ref: '#/components/parameters/cdi-version'
+        - $ref: '#/components/parameters/userId'
+      responses:
+        '200':
+          description: 'The operation was successful'
+          content:
+            application/json:
+                schema: 
+                  type: object
+                  properties:
+                    status:
+                      $ref: '#/components/schemas/statusOK'
+                    metadata:
+                      $ref: '#/components/schemas/userMetadata'
+        '400':
+          $ref: '#/components/responses/400'
+        
+        '401':
+          $ref: '#/components/responses/401'
+        
+        '404':
+          $ref: '#/components/responses/404'
+        
+        '500':
+          $ref: '#/components/responses/500'
+
+    post:
+      tags:
+        - User Metadata Recipe
+      operationId: userMetadataUpdate
+      description: |
+        Updates the metadata object stored about the user by doing a shallow merge of the stored and the update JSONs
+        and removing properties set to null on the root level of the update object. The merged object is then reserialized and stored.
```

Comment:
This seems to be more of a PUT API thing. Maybe change POST to PUT here?

Also, in the session recipe, we don't allow access token to be "updated" this way, instead, it just overrides whatever is there at the moment. Do you want to maintain that semantic? This implies that an update from the user's end would be first to fetch the metadata, then update it in mem, then query this API.

If instead, we want to maintain an update semantic, we should have a PUT API instead of POST, and we should be able to do a "deep" update. Else it can have unintended side effects from the user's point of view? 

On a related note, we don't need to have a POST API at all since an empty metadata object exists for all users anyway (even if that userId's row doesn't exist in the db).

Reply:
This can be PUT as well, it just looked like we used post for both updating and creating things (even delete, regenerate session, etc).

My reasoning behind the shallow merge was that you don't necessarily want to fetch the data unnecessarily, also, there is the edge-case of the parallel modifications.
Objects stored inside the session are different in my opinion since they are expected to be smaller/transient.

This merging mechanism (and objects in the root level being different) is a bit more complex but more convenient interface in my opinion. It also lessens the impact of the parallel modification edge case. We can solve that as well using ETags for example, but I'm not sure it's necessary.

I'm fine with going with full object overwrite if that's your preference.

Reply:
Im not sure what the best option is. What do other providers do for this?

Reply:
[Auth0](https://auth0.com/docs/api/management/v2#!/Users/patch_users_by_id):
Uses `PATCH` with shallow merging similar to what I described:

> The metadata fields are an exception to this rule (user_metadata and app_metadata). These properties are merged instead of being replaced but be careful, the merge only occurs on the first level.

Reply:
Okay. Let's stick with shallow merging then, but I would still prefer a PUT API method.